src/charm.py

```diff
@@ -124,6 +133,54 @@ def __init__(self, *args):
             self._certificates.on.certificate_expiring, self._on_certificate_expiring
         )
 
+    def _is_unit_in_non_active_status(self) -> Optional[StatusBase]:  # noqa: C901
+        """Check the possible conditions that Unit status is different than the Active.
+
+        Returns the Unit status that needs to be set if any condition is match, otherwise
+        returns None.
+
+        Returns:
+            StatusBase: MaintenanceStatus/BlockedStatus/WaitingStatus
+            None: If none of the conditionals match
+
+        """
+        if not self._amf_container.can_connect():
+            return MaintenanceStatus("Waiting for service to start")
+
+        if invalid_configs := self._get_invalid_configs():
+            return BlockedStatus(f"The following configurations are not valid: {invalid_configs}")
+
+        for relation in ["fiveg-nrf", "database", "certificates"]:
+            if not self._relation_created(relation):
+                return BlockedStatus(f"Waiting for {relation} relation")
+
+        if not self._database_is_available():
+            return WaitingStatus("Waiting for the amf database to be available")
+
+        if not self._get_database_info():
+            return WaitingStatus("Waiting for AMF database info to be available")
+
+        if not self._nrf_requires.nrf_url:
+            return WaitingStatus("Waiting for NRF data to be available")
+
+        if not self._amf_container.exists(path=CONFIG_DIR_PATH):
+            return WaitingStatus("Waiting for storage to be attached")
+
+        if not _get_pod_ip():
+            return WaitingStatus("Waiting for pod IP address to be available")
+
+        return None
+
+    def _on_collect_unit_status(self, event: CollectStatusEvent):
+        """Check the unit status and set to Unit when CollectStatusEvent is fired.
+
+        Args:
+            event: CollectStatusEvent
+        """
+        if self._is_unit_in_non_active_status():
+            event.add_status(self._is_unit_in_non_active_status())  # type: ignore[arg-type]
+            return
```

Comment:
Should we prefer event.add_status here ? According to [documentation](https://ops.readthedocs.io/en/latest/#ops.CollectStatusEvent) it says "Each code path in a collect-status handler should call add_status at least once."

Reply:
If we use ActiveStatus in else block, we lose the states which are set by functions triggered by relation_broken events. 
So, I am not setting ActiveStatus() through that hook. There is a strange situation. After a relation is broken, although we are checking the relations existence in def _is_unit_in_non_active_status, it could not catch it on time.

Reply:
I wonder if the issue you're seeing is related to https://github.com/canonical/operator/issues/940 or https://github.com/canonical/operator/issues/888? (Which we're hoping to address this cycle -- probably earlyish in the cycle.)

Yeah, ideally if you're using collect-status you'd avoid setting blocked in the relation broken handlers and just query for the status of the relation. But perhaps due to one of the above issues that's not working? What code would you be using to check for the relation's existence?

Reply:
Oh, yes, and to respond to your earlier comment -- sorry, that's my mistake! We should use `event.add_status(...)` here, oops!

Reply:
https://github.com/canonical/operator/issues/940 seems more related. I hope it will be implemented soon. 
What code would you be using to check for the relation's existence?
relations = [relation for relation in self.model.relations[relation_name]]
for rel in self._get_active_relations("fiveg-nrf"):
  # Try to find something that shows it does not active now. 
I really need your advice here.  
Please suggest me how to detect a broken relation.
Thank you!

Reply:
Yeah, that's a good question. Until #940 is addresses (which might be a little while, so don't hold your breath) you'll probably have to use a workaround. One workaround would be to set a variable/flag on the charm instance in the relation-broken hook, and then check that flag in the collect-status handler. Something like this (completely untested):

```python
class MyCharm(ops.CharmBase):
    def __init__(self, *args):
        super().__init__(*args)
        self.framework.observe(self.on.database_relation_broken, self._on_database_relation_broken)
        self.framework.observe(self.on.collect_unit_status, self._on_collect_unit_status)
        self._database_relation_breaking = False

    def _on_database_relation_broken(self, event):
        self._database_relation_breaking = True

    def _on_collect_unit_status(self, event):
        ...
        if not self.get_relation('database') or self._database_relation_breaking:
            event.add_status(ops.BlockedStatus('blah blah blah'))
            return
        ...
```

Reply:
What about storing a key similar in peer relation data similar to that example: https://github.com/canonical/postgresql-operator/blob/e13b871022bb6cff44cfb3ca83427c0e9d03f606/src/relations/db.py#L198C1-L202C19?

Reply:
I updated the PR adding a key into relation data. Unit tests are all fine but it is still failing in integration tests while detecting broken relation. Could you please have a look ?

Reply:
I added it quickly just to show you the implementation. There could be some improvement areas.

Reply:
Implemented according to your suggestion. Ready to review.